### PR TITLE
Various changes for DQ2 to work

### PR DESCRIPTION
**Pull request by timmysilv**
**Context:**
In #4436, I'm working on a mega-fix PR with far too many things to be reviewed at once. That said, CI is now passing there with most tests using the new device. I've copied over all changes to `pennylane/` to this fresh PR, and if CI passes here then we should be well on our way to swap out the device.

**Description of the Change:**
A lot of little things, but I'll try to sum them up here:
- if a device doesn't provide a `pennylane_requires`, assume it is always supported
- use observable data (as well as the state) to determine if we should use "backprop mode". If you have a circuit with non-trainable ops, but the observable has trainable data, the state will still be a non-trainable dtype at that point, and we might wrongly do CSR dot-product instead of sum of terms.
- don't support trainable Pow ops because they don't work in backprop (same as DQL)
- Squeeze all `CountsMP` results, not just dict results, to support batched CountsMP results
- Use `device.preprocess` instead of changing the expansion strategy for DQ2 in the `draw` module
- When sampling, only squeeze the last dimension with Pauli words to properly support `1xN` ShotCopies
- don't validate jitted DiagonalQubitUnitary
- `metric_tensor` has better interface handling so it can be multiplied by vanilla numpy arrays without breaking (needed for DQ2)
- QCut supports the new device interface

**Benefits:**
This PR is less than 200 lines and contains a significant portion of the little things needed to make DQ2 work as the new default.qubit

**Possible Drawbacks:**
Lots of unrelated changes, might be tough to review.

### CODE REVIEW

File: pennylane/interfaces/jax_jit.py [hunk 1]
```diff
@@ -29,11 +29,17 @@
...
+def _numeric_type_to_dtype(numeric_type, device):
     """Auxiliary function for converting from Python numeric types to JAX
     dtypes based on the precision defined for the interface."""
 
     single_precision = dtype is jnp.float32
+    if numeric_type is bool:
+        if isinstance(device, qml.Device):
+            numeric_type = int
+        else:
+            return jnp.bool_
```
Comment (albi3ro) [inline pennylane/interfaces/jax_jit.py:41]: This will go away if we start sampling in `int` instead of `bool`.

File: pennylane/qinfo/transforms.py
Comment (albi3ro) [inline pennylane/qinfo/transforms.py:1]: May have merge conflicts with story about qinfo transforms getting updated to new system.

File: pennylane/interfaces/execution.py [hunk 2]
```diff
@@ -254,6 +254,7 @@ def inner_execute(tapes: Sequence[QuantumTape], **_) -> ResultBatch:
             tapes = tuple(qml.transforms.convert_to_numpy_parameters(t) for t in tapes)
         return cached_device_execution(tapes)
 
+    # TODO: set as pure_callback if interface is jax-jit
```
Comment (albi3ro) [inline pennylane/interfaces/execution.py:257]: How does this tie into the device change?

Comment (timmysilv) [reply pennylane/interfaces/execution.py:257]: tbh I don't remember and it doesn't seem to be a problem anymore. I bet I wrote this before realizing something about ParametrizedEvolution etc. and it actually doesn't apply. deleting this TODO

File: pennylane/interfaces/jax_jit.py (hunk 1 above)
Comment (timmysilv) [reply pennylane/interfaces/jax_jit.py]: done

**Review by lillian542 — APPROVED** [review]
🚀

File: pennylane/tape/qscript.py [hunk 3]
```diff
@@ -793,11 +793,9 @@ def shape(self, device):
         if isinstance(device, qml.devices.experimental.Device):
             # MP.shape (called below) takes 2 arguments: `device` and `shots`.
             # With the new device interface, shots are stored on tapes rather than the device
-            # As well, MP.shape needs the device largely to see the device wires, and this is
-            # also stored on tapes in the new device interface. TODO: refactor MP.shape to accept
-            # `wires` instead of device (not currently done because probs.shape uses device.cutoff)
+            # TODO: refactor MP.shape to accept `wires` instead of device (not currently done
+            # because probs.shape uses device.cutoff)
```
Comment (mudit2812) [inline pennylane/tape/qscript.py:797]: This has bothered me for some time.

**Review by mudit2812 — APPROVED** [review]

File: pennylane/drawer/draw.py [hunk 4]
```diff
@@ -240,15 +240,23 @@ def _draw_qnode(
...
-        finally:
-            qnode.expansion_strategy = original_expansion_strategy
-
-        _wire_order = wire_order or getattr(qnode.device, "wires", None)
+        if expansion_strategy == "device" and isinstance(
+            qnode.device, qml.devices.experimental.Device
+        ):
+            qnode.construct(args, kwargs)
+            program, _ = qnode.device.preprocess()
+            tapes = program([qnode.tape])
```
[the remark this answers is not in the document]
Comment (timmysilv) [reply pennylane/drawer/draw.py:248]: it should be... but we iterate over `tapes[0]` below : ) batch-transform-moment

**Pull request merged**